packages/browser/src/client/tester/utils.ts

```diff
@@ -0,0 +1,180 @@
+      hasShadowRoot = true
+    }
+
+    const tag = el.tagName
+    if (el.id) {
+      path.push(`#${escapeIdForCSSSelector(el.id)}`)
+    }
+    else if (!el.nextElementSibling && !el.previousElementSibling) {
+      path.push(tag.toLowerCase())
+    }
+    else {
+      let index = 0
+      let sameTagSiblings = 0
+      let elementIndex = 0
+
+      for (const sibling of parent.children) {
+        index++
+        if (sibling.tagName === tag) {
+          sameTagSiblings++
+        }
+        if (sibling === el) {
+          elementIndex = index
+        }
+      }
+
+      if (sameTagSiblings > 1) {
+        path.push(`${tag.toLowerCase()}:nth-child(${elementIndex})`)
+      }
+      else {
+        path.push(tag.toLowerCase())
+      }
+    }
+    el = parent as Element
+  };
+  return `${getBrowserState().provider === 'webdriverio' && hasShadowRoot ? '>>>' : ''}${path.reverse().join(' > ')}`
+}
+
+function getParent(el: Element) {
+  const parent = el.parentNode
+  if (parent instanceof ShadowRoot) {
+    return parent.host
+  }
+  return parent
+}
+
+export class CommandsManager {
+  private _listeners: ((command: string, args: any[]) => void)[] = []
+
+  public onCommand(listener: (command: string, args: any[]) => void): void {
+    this._listeners.push(listener)
+  }
+
+  public async triggerCommand<T>(
+    command: string,
+    args: any[],
+    // error makes sure the stack trace is correct on webkit,
+    // if we make the error here, it looses the context
+    clientError: Error = new Error('empty'),
+  ): Promise<T> {
+    const state = getWorkerState()
+    const rpc = state.rpc as any as BrowserRPC
+    const { sessionId } = getBrowserState()
+    const filepath = state.filepath || state.current?.file?.filepath
+    args = args.filter(arg => arg !== undefined) // remove optional fields
+    if (this._listeners.length) {
+      await Promise.all(this._listeners.map(listener => listener(command, args)))
+    }
+    return rpc.triggerCommand<T>(sessionId, command, filepath, args).catch((err) => {
+      // rethrow an error to keep the stack trace in browser
+      // const clientError = new Error(err.message)
+      clientError.message = err.message
+      clientError.name = err.name
+      clientError.stack = clientError.stack?.replace(clientError.message, err.message)
+      throw clientError
+    })
+  }
+}
+
+const now = Date.now
+
+export function processTimeoutOptions<T extends { timeout?: number }>(options_?: T): T | undefined {
+  if (
+    // if timeout is set, keep it
+    (options_ && options_.timeout != null)
+    // timeout can only be set for playwright commands
+    || provider !== 'playwright'
+  ) {
+    return options_
+  }
+  // if there is a default action timeout, use it
+  if (getWorkerState().config.browser.providerOptions.actionTimeout != null) {
+    return options_
+  }
+  const currentTest = getWorkerState().current
+  const startTime = currentTest?.result?.startTime
```

Comment:
This might not be a good idea. The `current` will return the test even if it runs `beforeEach` and `afterEach`, and the timeout there must be different!